name.abuchen.portfolio/src/name/abuchen/portfolio/online/impl/HTMLTableQuoteFeed.java

```diff
@@ -308,16 +315,64 @@ public boolean updateHistoricalQuotes(Security security, List<Exception> errors)
             return Collections.emptyList();
         }
 
-        List<LatestSecurityPrice> answer = cache.lookup(feedURL);
-        if (answer != null)
-            return answer;
+        VariableURL variableURL;
 
```

Comment:
Im Prinzip ist jetzt die Logik für die beiden verfügbaren Platzhalter in die ```internalGetQuotes``` Methode kodiert. Und zwar genau in dieser Reihenfolge: erst das Datum, dann die Page. Das macht diese Methode nicht nur unnötig kompliziert, sondern macht es auch schwerer weitere Platzhalter in Zukunft hinzuzufügen. Und die Reihenfolge ist hartkodiert.

Mal so als Brainstorming: Die ```variableurl.Factory``` kennt doch die verschiedene Platzhalter. Und jeder Platzhalter sollte seine "natürliche" Iteration kennen (also PAGE von 1 bis unendlich; DATE von "last price" bis "now"). Dann ist die allgemeine Logik: pro "Iterator" einen Schleife.

Damit das geht, müsste man
a) eine Reihenfolge angeben können (erst Datum, dann Page)
b) eine gemeinsame Abbruchbedingung haben (aktuell sind es mal 100 Tage ohne neue Kurse bzw. eine Seite ohne neue Kurse)

Was hälst Du von so einem Vorschlag:

* Die Platzhalter haben immer folgendes Format: ```%{order:name:format}``` also z.B. ```%{0:DATE:yyyy-MM}``` und ```%{1:PAGE}```
* Die Abbruchbedingung ist immer: wenn eine Query keine neuen Kurse bringt, dann wird der Iterator abgebrochen (damit Dein Beispiel funktioniert, müsste man einen ```%{0:MONTHLY:yyyy-MM-32``` hinzufügen)

Reply:
Ehrlich gesagt habe ich %{PAGE} nur "aus Prinzip" eingebaut. Es bietet mit genau 0 Mehrwert, aber es könnte ja einer daherkommen, dessen Kurslieferant...

Gegenvorschlag: entweder %{DATE:yyyy-MM} oder %{PAGE}, aber nicht beides. Das würde so einiges vereinfachen.

Reply:
> Gegenvorschlag: entweder %{DATE:yyyy-MM} oder %{PAGE}, aber nicht beides. Das würde so einiges vereinfachen.

Wenn ich das richtig sehe, ist das Makro PAGE weiterhin drin. Das passt doch. Ich werde für meine URLs noch was wie {NOW:yyyy-MM-dd:-1y} also das aktuelle Datum minus ein Jahr.

Reply:
Nein, ich meine *pro URL* soll man nur {DATE:...} oder {PAGE} angeben können. Einverstanden?

{NOW:yyyy-MM-dd:-1y} gefällt mir insofern nicht als dass sich der User IMO nicht darum kümmern müssen sollte, welches Datum denn genau drin steht. Wenn ich bestimmte Spannen brauche, kann ich ja sowas machen: `from={DATE:yyyy-MM-01}&to={DATE:yyyy-MM-32}`

Reply:
> {NOW:yyyy-MM-dd:-1y} gefällt mir insofern nicht als dass sich der User IMO nicht darum kümmern müssen sollte, welches Datum denn genau drin steht.

Kommt drauf an. Bei anderen Seiten sieht die URL ungefähr so aus: ```....?start=03.06.2016&range=1yr```. Ich ändere das Datum dann auf gestern und habe so ca. 1 Jahr Ruhe. Dann muss ich die URL wieder ändern. In diesem Fall würde ich gerne ein relatives Datum angeben wie "Heute - 1 Jahr".

Wie auch immer, das sollte nicht in diesen Change rein.

Reply:
1. Bevor wir im Puncto DATE xor PAGE weiter Ping-Pong spielen, baue ich das einfach mal um und dann kömma immer noch schauen.
2. `...?start={DATE:01.01.yyyy}&range=1yr` ?

Reply:
@buchen Sodale! Jetzt ist das ganze doch um einiges übersichtlicher. Jede Variation der URL weiß, welchen Weg sie zu gehen hat und stellt eigene `maxFailedAttempts`.

Reply:
> Jetzt ist das ganze doch um einiges übersichtlicher.

Stimmt. Mir gefällt auch besser, das die Logik um doppelte URLs auszusortieren in den Iterator gewandert ist. 

Wenn ich das richtig sehe sind die "maxFailedAttempts" ebenfalls dynamisch. Bei "Monaten" (das ist ja Dein Beispiel) ist man wohl bei 100 / (366/12) ~ 3 weiteren Versuchen. D.h. wenn die Daten bis April zurückreichen, dann werden noch Daten angefragt für den März, Februar und Januar abgefragt. Ich denke das ist okay. Das kann man ggf. auch später tunen.

Ich nehme die Change mit rein. Vermutlich nächstes Wochenende werde ich eine neue Version bauen und veröffentlichen. Willst Du noch Unit Tests in diesen PR pushen? Oder eher einen neuen PR aufmachen?

Reply:
Keine Ahnung ob die Deadline jetzt 9. oder 16. ist (soll mir gleich sein) – die U.T. gibts hier noch mit dazu.

Reply:
Und... bitteschön.